src/shadowbox/server/server_access_key.spec.ts

```diff
@@ -92,8 +94,102 @@ describe('ServerAccessKeyRepository', () => {
     });
   });
 
+  it('Creates keys at the right port by construction', async (done) => {
+    const portProvider = new PortProvider();
+    const port = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().port(port).build();
+    const key = await repo.createNewAccessKey();
+    expect(key.proxyParams.portNumber).toEqual(port);
+    done();
+  });
+
+  it('setDefaultPort changes default port for new keys', async (done) => {
+    const portProvider = new PortProvider();
+    const port = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().build();
+    await repo.setDefaultPort(port);
+    const key = await repo.createNewAccessKey();
+    expect(key.proxyParams.portNumber).toEqual(port);
+    done();
+  });
+
+  it('setDefaultPort maintains ports on existing keys', async (done) => {
+    const portProvider = new PortProvider();
+    const oldPort = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().port(oldPort).build();
+    const oldKey = await repo.createNewAccessKey();
+
+    const newPort = await portProvider.reserveNewPort();
+    await repo.setDefaultPort(newPort);
+    expect(oldKey.proxyParams.portNumber).toEqual(oldPort);
+    done();
+  });
+
+  it('setDefaultPort rejects invalid port numbers', async (done) => {
+    const repo = new RepoBuilder().build();
+    // jasmine.toThrowError expects a function and makes the code
+    // hard to read.
+    const expectThrow = async (port: number) => {
+      try {
```

Comment:
I like the wrapper for this test. Maybe you can further simplify it by replacing the try-catch with:  `expect(await repo.setDefaultPort(port)).toThrowError(errors.InvalidPortNumber)`.

Reply:
See above